File: lib/core/models/route_error.dart
import 'dart:async';
import 'dart:io';

import 'package:http/http.dart' as http;

import '../../l10n/app_localizations.dart';
import '../services/route_service.dart';

/// ルート計算失敗の種別。UI の文言と復帰導線の出し分けに使う。
enum RouteErrorKind {
  network,
  timeout,
  noResults,
  noLocation,
  noDestination,
  unknown,
}

/// エラー画面で主に提示する復帰アクション。
enum RouteRecovery { retry, changeConditions }

class RouteErrorView {
  const RouteErrorView({
    required this.kind,
    required this.title,
    required this.description,
    required this.primaryRecovery,
  });

  final RouteErrorKind kind;
  final String title;
  final String description;
  final RouteRecovery primaryRecovery;
}

/// 例外をユーザー向けエラー種別へ分類する。
RouteErrorKind classifyRouteError(Object error) {
  if (error is RouteException) {
    switch (error.status) {
      case 'NO_ORIGIN':
        return RouteErrorKind.noLocation;
      case 'NO_DESTINATION':
        return RouteErrorKind.noDestination;
      case 'ZERO_RESULTS':
      case 'NOT_FOUND':
        return RouteErrorKind.noResults;
      case 'UNKNOWN_ERROR':
      case 'UNKNOWN':
      case 'OVER_QUERY_LIMIT':
        return RouteErrorKind.network;
      // TIMEOUT: サービス最下層が TimeoutException を変換したもの（#156）と、検索の
      // 締切による打ち切り（#300）。network に丸めない——上流 Transit API の遅延は
      // 通信断ではなく、「通信状況を確認して再試行」は電波が正常なユーザーを的外れな
      // 導線へ送る。#300 の切り分けでは、画面表示から App Check 拒否・レート制限・
      // 上流遅延を区別できないこと自体が障害になった。
      case 'TIMEOUT':
        return RouteErrorKind.timeout;
    }
    if (error.status.startsWith('HTTP')) return RouteErrorKind.network;
    // NO_PROXY / REQUEST_DENIED / INVALID_REQUEST など設定・権限系。
    return RouteErrorKind.unknown;
  }
  // 素の TimeoutException（ドメイン例外へ変換される前に漏れたもの）も遅延として扱う。
  if (error is TimeoutException) return RouteErrorKind.timeout;
  // SocketException / HttpException / HandshakeException など dart:io 系、
  // http パッケージの ClientException は通信系として扱う。
  if (error is IOException || error is http.ClientException) {
    return RouteErrorKind.network;
  }
  return RouteErrorKind.unknown;
}

RouteErrorView routeErrorView(AppLocalizations l10n, RouteErrorKind kind) =>
    switch (kind) {
      RouteErrorKind.network => RouteErrorView(
        kind: RouteErrorKind.network,
        title: l10n.routeErrorNetworkTitle,
        description: l10n.routeErrorNetworkDescription,
        primaryRecovery: RouteRecovery.retry,
      ),
      RouteErrorKind.timeout => RouteErrorView(
        kind: RouteErrorKind.timeout,
        title: l10n.routeErrorTimeoutTitle,
        description: l10n.routeErrorTimeoutDescription,
        primaryRecovery: RouteRecovery.retry,
      ),
      RouteErrorKind.noResults => RouteErrorView(
        kind: RouteErrorKind.noResults,
        title: l10n.routeErrorNoResultsTitle,
        description: l10n.routeErrorNoResultsDescription,
        primaryRecovery: RouteRecovery.changeConditions,
      ),
      RouteErrorKind.noLocation => RouteErrorView(
        kind: RouteErrorKind.noLocation,
        title: l10n.routeErrorNoLocationTitle,
        description: l10n.routeErrorNoLocationDescription,
        primaryRecovery: RouteRecovery.retry,
      ),
      RouteErrorKind.noDestination => RouteErrorView(
        kind: RouteErrorKind.noDestination,
        title: l10n.routeErrorNoDestinationTitle,
        description: l10n.routeErrorNoDestinationDescription,
        primaryRecovery: RouteRecovery.changeConditions,
      ),
      RouteErrorKind.unknown => RouteErrorView(
        kind: RouteErrorKind.unknown,
        title: l10n.routeErrorUnknownTitle,
        description: l10n.routeErrorUnknownDescription,
        primaryRecovery: RouteRecovery.retry,
      ),
    };

File: apps/web/src/state/route-error.ts
// 移植元: lib/core/models/route_error.dart
//
// 文言と復帰導線と対で意味を持つので、エラー画面のスライスまで運んでいなかったもの
// （PORTING.md の「まだ運んでいないもの」）。

import {
  ClientException,
  TimeoutException,
} from '@aruku/engine/services/http-client';
import { RouteException } from '@aruku/engine/services/route-service';

import { ja } from '../i18n/ja';
import { RouteErrorKind } from './app-state';

/// エラー画面で主に提示する復帰アクション。
export const RouteRecovery = {
  retry: 'retry',
  changeConditions: 'changeConditions',
} as const;
export type RouteRecovery = (typeof RouteRecovery)[keyof typeof RouteRecovery];

export interface RouteErrorView {
  readonly kind: RouteErrorKind;
  readonly title: string;
  readonly description: string;
  readonly primaryRecovery: RouteRecovery;
}

/// エンジンが投げる `RouteException` のステータス → ユーザー向け種別。
///
/// 移植元にあった `NOT_FOUND` / `UNKNOWN_ERROR` / `UNKNOWN` / `OVER_QUERY_LIMIT` は
/// 運んでいない。NAVITIME 時代の名残で、#330 の撤去以降は Dart のサービス層も
/// これらを投げない（実際に投げているのは下の6つと `HTTP <code>` だけ）。
const kindByStatus: Readonly<Record<string, RouteErrorKind>> = {
  NO_ORIGIN: RouteErrorKind.noLocation,
  NO_DESTINATION: RouteErrorKind.noDestination,
  ZERO_RESULTS: RouteErrorKind.noResults,
  // 上流の遅延は通信断ではない。network へ丸めると「通信状況を確認して再試行」が
  // 電波の正常なユーザーを的外れな導線へ送る（#300 の切り分けで障害になった点）。
  TIMEOUT: RouteErrorKind.timeout,
  // 設定・配線の失敗。ユーザーの操作では直らないので、再試行も条件変更も勧めない。
  NO_TRANSIT_API: RouteErrorKind.unknown,
  MATRIX_NOT_ARRAY: RouteErrorKind.unknown,
};

/// 例外をユーザー向けエラー種別へ分類する。
export function classifyRouteError(error: unknown): RouteErrorKind {
  if (error instanceof RouteException) {
    const known = kindByStatus[error.status];
    if (known !== undefined) return known;
    if (error.status.startsWith('HTTP')) return RouteErrorKind.network;
    return RouteErrorKind.unknown;
  }

  // ドメイン例外へ変換される前に漏れたタイムアウト。
  if (error instanceof TimeoutException) return RouteErrorKind.timeout;

  // 通信の失敗はトランスポート層が ClientException へ寄せて届く
  // （fetch-http-client.ts の `asClientException`。`TypeError: Failed to fetch` も含む）。
  //
  // 移植元は dart:io の IOException と http の ClientException を**両方**見ていた。
  // dart2js の dart:io がスタブで Web の通信断が IOException に一致しないため（#359）。
  // こちらは層が1つに揃えているので、その二重性は運ばない。
  //
  // 裏返しとして、**素の TypeError をここで拾ってはいけない**。通信由来のものは上記の
  // とおり包まれて届くので、ここへ来る TypeError は我々のバグ——network に寄せると
  // 「通信状況を確認してください」として、ユーザーにも我々にも見えなくなる。
  if (error instanceof ClientException) return RouteErrorKind.network;

  return RouteErrorKind.unknown;
}

const viewByKind: Readonly<Record<RouteErrorKind, Omit<RouteErrorView, 'kind'>>> = {
  [RouteErrorKind.network]: {
    title: ja.routeErrorNetworkTitle,
    description: ja.routeErrorNetworkDescription,
    primaryRecovery: RouteRecovery.retry,
  },
  [RouteErrorKind.timeout]: {
    title: ja.routeErrorTimeoutTitle,
    description: ja.routeErrorTimeoutDescription,
    primaryRecovery: RouteRecovery.retry,
  },
  [RouteErrorKind.noResults]: {
    title: ja.routeErrorNoResultsTitle,
    description: ja.routeErrorNoResultsDescription,
    // 同じ条件で再試行しても同じ結果になる。勧めるのは条件の変更。
    primaryRecovery: RouteRecovery.changeConditions,
  },
  [RouteErrorKind.noLocation]: {
    title: ja.routeErrorNoLocationTitle,
    description: ja.routeErrorNoLocationDescription,
    primaryRecovery: RouteRecovery.retry,
  },
  [RouteErrorKind.noDestination]: {
    title: ja.routeErrorNoDestinationTitle,
    description: ja.routeErrorNoDestinationDescription,
    primaryRecovery: RouteRecovery.changeConditions,
  },
  [RouteErrorKind.unknown]: {
    title: ja.routeErrorUnknownTitle,
    description: ja.routeErrorUnknownDescription,
    primaryRecovery: RouteRecovery.retry,
  },
};

export function routeErrorView(kind: RouteErrorKind): RouteErrorView {
  return { kind, ...viewByKind[kind] };
}
